core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClientWithInvokerAgent.scala

```diff
@@ -0,0 +1,147 @@
+ * instances running on every worker node that runs user containers to provide
+ * suspend/resume capability and higher performance log processing.
+ */
+class KubernetesClientWithInvokerAgent(config: KubernetesClientConfig)(executionContext: ExecutionContext)(
+  implicit log: Logging,
+  as: ActorSystem)
+    extends KubernetesClient(config)(executionContext)
+    with KubernetesApiWithInvokerAgent {
+
+  override def rm(key: String, value: String, ensureUnpaused: Boolean = false)(
+    implicit transid: TransactionId): Future[Unit] = {
+    if (ensureUnpaused) {
+      // The caller can't guarantee that every container with the label key=value is already unpaused.
+      // Therefore we must enumerate them and ensure they are unpaused before we attempt to delete them.
+      Future {
+        blocking {
+          kubeRestClient
+            .inNamespace(kubeRestClient.getNamespace)
+            .pods()
+            .withLabel(key, value)
+            .list()
+            .getItems
+            .asScala
+            .map { pod =>
+              val container = toContainer(pod)
+              container
+                .resume()
+                .recover { case _ => () } // Ignore errors; it is possible the container was not actually suspended.
+                .map(_ => rm(container))
+            }
+        }
+      }.flatMap(futures =>
+        Future
+          .sequence(futures)
+          .map(_ => ()))
+    } else {
+      super.rm(key, value, ensureUnpaused)
+    }
+  }
+
+  override def suspend(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
+    agentCommand("suspend", container)
+      .map { response =>
+        response.discardEntityBytes()
+      }
+  }
+
+  override def resume(container: KubernetesContainer)(implicit transid: TransactionId): Future[Unit] = {
+    agentCommand("resume", container)
+      .map { response =>
+        response.discardEntityBytes()
+      }
+  }
+
+  def forwardLogs(container: KubernetesContainer,
+                  lastOffset: Long,
+                  sizeLimit: ByteSize,
+                  sentinelledLogs: Boolean,
+                  additionalMetadata: Map[String, JsValue],
+                  augmentedActivation: JsObject)(implicit transid: TransactionId): Future[Long] = {
+    val serializedData = Map(
+      "lastOffset" -> JsNumber(lastOffset),
+      "sizeLimit" -> JsNumber(sizeLimit.toBytes),
+      "sentinelledLogs" -> JsBoolean(sentinelledLogs),
+      "encodedLogLineMetadata" -> JsString(fieldsString(additionalMetadata)),
+      "encodedActivation" -> JsString(augmentedActivation.compactPrint))
+
+    agentCommand("logs", container, serializedData)
+      .flatMap { response =>
+        Unmarshal(response.entity).to[String].map(s => s.toLong)
+      }
+  }
+
+  def agentCommand(command: String,
+                   container: KubernetesContainer,
+                   payload: Map[String, JsValue] = Map.empty): Future[HttpResponse] = {
```

Comment:
What do you think about changing the interface to take `Option[JsObject]` instead? Callers essentially build a JsObject anyway already and the Option would pass **no** entity down if not needed vs. passing an empty JsObject in all those cases.